import type { Client } from "../types";

const URL_API = "http://localhost:5070";
const myHeaders = new Headers({
    "Content-Type": "application/json"
});

export const getClient = async (): Promise<Client[] | null> => {
    try {
        const response = await fetch(`${URL_API}/api/Client`, {
            method: 'GET',
            headers: myHeaders
        });

        switch (response.status) {
            case 200:
                const data: Client[] = await response.json();
                return data;
            case 401:
                console.error("No autorizado o token inv√°lido");
                break;
            case 404:
                console.error("El cliente no existe");
                break;
            default:
                console.error("Error inesperado. Contacte al administrador.");
        }
    } catch (error) {
        console.error("Error de red o servidor:", error);
    }

    return null; // en caso de error
};

export const postClient = async (datos: Client): Promise<any | undefined> => {
    try {
        // Remove id if present
        const { id, ...clientData } = datos;
        console.log("Datos enviados a postClient:", clientData);

        const response = await fetch(`${URL_API}/api/Client`, {
            method: "POST",
            headers: myHeaders,
            body: JSON.stringify(clientData)
        });
        if (!response.ok) {
            const errorText = await response.text();
            console.error(`Error en la solicitud POST: ${response.status} - ${errorText}`);
            return undefined;
        }
        return await response.json();
    } catch (error) {
        console.error('Error en la solicitud POST:', error);
    }
}

export const putClient = async (datos: Client, id: number | string): Promise<Response | undefined> => {
    try {
        return await fetch(`${URL_API}/api/Client/${id}`, {
            method: "PUT",
            headers: myHeaders,
            body: JSON.stringify(datos)
        });
    } catch (error) {
        console.error('Error en la solicitud PUT:', error);
    }
}

export const deleteClient = async (id: number | string): Promise<Response | undefined> => {
    try {
        const response = await fetch(`${URL_API}/api/Client/${id}`, {
            method: "DELETE",
            headers: myHeaders,
        });
        if (!response.ok) {
            const errorText = await response.text();
            console.error(`Error en la solicitud DELETE: ${response.status} - ${errorText}`);
        }
        return response;
    } catch (error) {
        console.error('Error en la solicitud DELETE:', error);
    }
}